crates/turbo-tasks-memory/src/task.rs

```diff
@@ -672,24 +706,44 @@ impl Task {
     /// Prepares task execution and returns a future that will execute the task.
     fn make_execution_future(
         self: &Task,
-        mut state: &mut TaskState,
+        mut state: FullTaskWriteGuard<'_>,
+        backend: &MemoryBackend,
         turbo_tasks: &dyn TurboTasksBackendApi,
     ) -> Pin<Box<dyn Future<Output = Result<RawVc>> + Send>> {
         match &self.ty {
-            TaskType::Root(bound_fn) => bound_fn(),
-            TaskType::Once(mutex) => mutex.lock().take().expect("Task can only be executed once"),
+            TaskType::Root(bound_fn) => {
+                drop(state);
+                bound_fn()
+            }
+            TaskType::Once(mutex) => {
+                drop(state);
+                mutex.lock().take().expect("Task can only be executed once")
+            }
+            &TaskType::ReadCollectibles(scope_id, trait_type_id) => {
+                drop(state);
+                Task::make_read_collectibles_execution_future(
+                    self.id,
+                    scope_id,
+                    trait_type_id,
+                    backend,
+                    turbo_tasks,
+                )
+            }
             TaskType::Persistent(ty) => match &**ty {
                 PersistentTaskType::Native(native_fn, inputs) => {
-                    if let PrepareTaskType::Native(bound_fn) = &state.prepared_type {
+                    let future = if let PrepareTaskType::Native(bound_fn) = &state.prepared_type {
                         bound_fn()
```

Comment:
Why not drop state right before caling `bound_fn()` like we do for `::Root` above?

Reply:
bound_fn is borrowed from state, so we can't drop it earlier